### PR TITLE
Stat1

### PR DESCRIPTION
**Pull request by ahtavangar**

### CODE REVIEW

**Comment by ahtavangar** [comment]
Hi David,
I have added ANOVA macros to stats library. Please let me know what you think.

Amir

**Comment by drlippman** [comment]
Thanks for this.  Unfortunately your PR also includes several other files.  When making a PR, it's best to create a new branch off the original master, so that any changes are separate and distinct.  For now, I'll look over these stats functions, and can merge them manually.

File: assessment/libs/stats.php [hunk 1]
```diff
@@ -2196,4 +2197,393 @@ function dotplot($a,$label,$dotspace=1,$labelspace=null,$width=300,$height=150)
...
+			return false;
+		}
+	}
+	if (count($n)<2) {
+		echo "Error: ANOVA requires at least two arrays";
+		return false;
+	}
+		
+	$numargs = func_num_args();
+	$args=func_get_args();
```
Comment (drlippman) [inline assessment/libs/stats.php:2226]: I'm confused why you're using the spread operator `... $arr` in the function declaration, but then using `func_get_args` here, when this is returning the same thing that $arr would contain.

File: assessment/libs/stats.php [hunk 2]
```diff
@@ -2196,4 +2197,393 @@ function dotplot($a,$label,$dotspace=1,$labelspace=null,$width=300,$height=150)
...
+	$mean=array();
+	$ss=array();
+	for($i=0;$i<$numargs;$i++){
+		$mean[$i]=mean($args[$i]);
+		$ss[$i]=variance($args[$i])*(count($args[$i])-1);
+		//$n[$i]=count($args[$i]);
+	}
+	$gmean=array_sum($mean)/count($mean); //grand mean
+	$ssA=variance($mean)*$n[0]*(count($args)-1); //Sum of the square for Factor A
```
Comment (drlippman) [inline assessment/libs/stats.php:2235]: Why use `count($args)` here when you already have `$numargs`?

File: assessment/libs/stats.php [hunk 3]
```diff
@@ -2196,4 +2197,393 @@ function dotplot($a,$label,$dotspace=1,$labelspace=null,$width=300,$height=150)
...
+		$p_a=fcdf($F_a,$dfA,$dfE); //P value of factor A
+		$p_b=fcdf($F_b,$dfB,$dfE); //P value of factor B
+		$ans=array([$ssA,$dfA,$msA,$F_a,$p_a],[$ssB,$dfB,$msB,$F_b,$p_b],[$ssE,$dfE,$msE],[$ssT,$dfT]);
+			
+	}
+	return ($ans);	
+}
+
+//---------------------------------------------ANOVA-Twoway F ratio-----------------------------------------
+// Function: anova1way_f(arr, [replication=False])
```
Comment (drlippman) [inline assessment/libs/stats.php:2485]: Not a biggie, but should be anova2way_f here

File: assessment/libs/stats.php [hunk 4]
```diff
@@ -2196,4 +2197,393 @@ function dotplot($a,$label,$dotspace=1,$labelspace=null,$width=300,$height=150)
...
+// NameA: Optional - the name of factor A as string to be displayed in the table. Default is "Factor A".
+// NameB: Optional - the name of factor B as string to be displayed in the table. Default is "Factor B".
+// Returns:
+// ANOVA table for displaying data. 
+
+
+function anova_table(array $array, int $factor = 1, $rep=False, int $roundto=12, string $f1="Factor A", string $f2="Factor B"){
+	if ($factor!=1 && $factor!=2) { echo 'error: the factor variable only expects 1 for one-way and 2 for two-way ANOVA'; return '';}
+	if (!function_exists('calconarray')) {
+        require_once(__DIR__.'/assessment/macros.php');
```
Comment (drlippman) [inline assessment/libs/stats.php:2539]: This shouldn't be necessary; macros.php is always loaded when the macro libraries are used.

File: assessment/libs/stats.php [hunk 5]
```diff
@@ -2196,4 +2197,393 @@ function dotplot($a,$label,$dotspace=1,$labelspace=null,$width=300,$height=150)
...
+      }
+	
+	if ($factor==1){
+		$r0=calconarray($array[0],"round(x,$roundto)");
```
Comment (drlippman) [inline assessment/libs/stats.php:2543]: It'd be slightly more efficient here to use a native PHP function.  You could round all the elements in the subarrays of $array using:

`array_walk_recursive($a, function(&$x) use ($roundto) { $x = round($x,$roundto);});`

**Comment by drlippman** [comment]
I'm happy to make the suggested changes and manually merge, if you'd like.  If you'd prefer to make the changes and submit an isolated PR, let me know.

**Pull request closed without merging**